float64 length
float64 smoothness
